Add `array` lang item and `[T; N]::map(f: FnMut(T) -> S)`

JulianKnodt opened the pull request:
This introduces an `array` lang item so functions can be defined on top of `[T; N]`. This was previously not done because const-generics was not complete enough to allow for this. Now it is in a state that is usable enough to start adding functions.

The function added is a monadic (I think?) map from `[T; N] -> [S; N]`. Until transmute can function on arrays, it also allocates an extra temporary array, but this can be removed at some point.

r? @lcnr

JulianKnodt committed:
Create lang item array and add map fn

This creates the language item for arrays, and adds the map fn which is like map in options or
iterators. It currently allocates an extra array, unfortunately.

Added fixme for transmuting

Fix typo

Add drop guard

c410-f3r commented on library/core/src/array/mod.rs:
```diff
@@ -364,3 +364,30 @@ macro_rules! array_impl_default {
+#[lang = "array"]
+#[cfg(not(bootstrap))]
+impl<T, const N: usize> [T; N] {
+    /// Returns an array of the same aize as self, with `f` applied to each element.
```
Typo

```suggestion
    /// Returns an array of the same size as self, with `f` applied to each element.
```

jonas-schievink commented on library/core/src/array/mod.rs:
```diff
@@ -364,3 +364,30 @@ macro_rules! array_impl_default {
+        use crate::mem::MaybeUninit;
+        let dst = MaybeUninit::uninit_array::<N>();
+        for (i, e) in self.into_iter().enumerate() {
+            dst[i] = MaybeUninit::new(f(e));
```
This looks like it would leak all previous elements if f panics

JulianKnodt replied:
Do you mean of `dst` or of the iterator?

jonas-schievink replied:
of `dst` (since `MaybeUninit` won't run the destructor of the contained value)

CryZe replied:
Of dest because it's MaybeUninit so nothing will get dropped automatically.

lcnr replied:
`dst`, see the implementation of `fn make` on how this can be handled:
https://github.com/rust-lang/rust/blob/c448b373b57d0f1adefb70e07a44b1cccc9d2393/library/core/src/array/mod.rs#L357-L385

cuviper commented on library/core/src/array/mod.rs:
```diff
@@ -364,3 +364,56 @@ macro_rules! array_impl_default {
+        }
+        // FIXME convert to crate::mem::transmute when works with generics
+        // unsafe { crate::mem::transmute::<[MaybeUninit<S>; N], [S; N]>(dst) }
+        unsafe { (&mut dst as *mut _ as *mut [S; N]).read() }
```
You'd better `forget(guard)` first.

JulianKnodt replied:
Would it be necessary to also manually drop `dst` afterwards?

cuviper replied:
What would you need to `drop`? `MaybeUninit` doesn't touch its contents, since it doesn't know their validity. You're moving the items out with the pointer read, so I think that's it.

JulianKnodt replied:
Ah I thought the `ptr.read()` creates a copy of the array, and doesn't move the items out

cuviper replied:
The only difference between a move and a copy is what happens to the original location. We don't access `dst` anymore, and `MaybeUninit` won't touch the data when it goes out of scope, so it's effectively moved.

JulianKnodt replied:
ah ok, I guess we also just don't want to run the destructors of the temporary array anyway, so this makes more sense

cuviper commented on library/core/src/array/mod.rs:
```diff
@@ -364,3 +364,56 @@ macro_rules! array_impl_default {
+        }
+        let dst = MaybeUninit::uninit_array::<N>();
+        let mut guard = Guard::new(&mut dst);
+        for (i, e) in self.into_iter().enumerate() {
```
Doesn't `self.into_iter()` still auto-deref to a slice iterator? I think you have to manually `IntoIter::new(self)` if you want to move.

JulianKnodt committed:
Add Array Impl Lang Item in various places

Add basic test

And also run fmt which is where the other changes are from

Fix mut issues

These only appear when running tests, so resolved by adding mut

Swap order of forget

Add pub and rm guard impl

Add explicit type to guard

Add safety note

Change guard type from T to S

It should never have been T, as it guards over [MaybeUninit<S>; N]
Also add feature to test

cuviper commented on library/core/src/array/mod.rs:
```diff
@@ -364,3 +364,53 @@ macro_rules! array_impl_default {
+            }
+        }
+        let mut dst = MaybeUninit::uninit_array::<N>();
+        let mut guard: Guard<T, N> = Guard { dst: &mut dst as *mut _ as *mut T, curr_init: 0 };
```
Shouldn't this be a `Guard<S, N>`? You're writing the result of `f(e)` into `dst`, and the guard will drop those partial results from `dst` if needed.

JulianKnodt replied:
woops yea

cuviper commented on library/core/src/array/mod.rs:
```diff
@@ -364,3 +364,53 @@ macro_rules! array_impl_default {
+        let mut dst = MaybeUninit::uninit_array::<N>();
+        let mut guard: Guard<T, N> = Guard { dst: &mut dst as *mut _ as *mut T, curr_init: 0 };
+        for (i, e) in IntoIter::new(self).enumerate() {
+            dst[i] = MaybeUninit::new(f(e));
```
I think this can be `dst[i].write(f(e))`. You could also `zip(&mut dst)` to avoid enumeration and indexing.

JulianKnodt replied:
Is this in general preferable? For const-generics I find the indexing to be elegant, but that's more of a personal preference

cuviper replied:
In general, I'd say yes, but I don't have enough experience with const-generics and how well they optimize.

JulianKnodt replied:
yea I'd hope the indexing would always be optimized away but I guess can't be sure.

JulianKnodt committed:
Add recommend changes to array

Switch from indexing to zip, and also use `write` on `MaybeUninit`.

Add array_map feature to core/src/lib

Attempt to fix issue of no such feature

Update w/ pickfire's review

This changes a couple of names around, adds another small test of variable size,
and hides the rustdoc #![feature(..)].

Fmt doctest

Add suggestions from lcnr

pickfire commented on library/core/src/array/mod.rs:
````diff
@@ -364,3 +364,54 @@ macro_rules! array_impl_default {
+    ///
+    /// # Examples
+    /// ```
+    /// #![feature(array_map)]
````
```suggestion
    /// # #![feature(array_map)]
```

Should we hide this? rustdoc already reports it.

pickfire commented on library/core/src/array/mod.rs:
```diff
@@ -364,3 +364,54 @@ macro_rules! array_impl_default {
+    #[unstable(feature = "array_map", issue = "77777")]
+    pub fn map<F, S>(self, mut f: F) -> [S; N]
+    where
+        F: FnMut(T) -> S,
```
Should this be `U` rather than `S`? Most of the other `map` uses `U` except `Iterator::map` which uses `B`.

JulianKnodt replied:
Ah no, I just picked S randomly, will change

pickfire commented on library/core/src/array/mod.rs:
```diff
@@ -364,3 +364,54 @@ macro_rules! array_impl_default {
+        use crate::mem::MaybeUninit;
+        struct Guard<T, const N: usize> {
+            dst: *mut T,
+            curr_init: usize,
```
Should we rename this to `len` or `initialized` to be clearer?

JulianKnodt replied:
I'm not sure if these are clearer; I find len to be much less clear, I'll change it to initialized because it does indicate that these values were initialized previously.

pickfire commented on library/core/tests/array.rs:
```diff
@@ -290,3 +290,10 @@ fn empty_array_is_always_default() {
+    let a = [1, 2, 3];
+    let b = a.map(|v| v + 1);
+    assert_eq!(b, [2, 3, 4]);
+}
```
Should we test converting this to another type with different size?

pickfire commented on library/core/src/array/mod.rs:
```diff
@@ -372,28 +372,28 @@ impl<T, const N: usize> [T; N] {
+    /// # #![feature(array_map)]
     /// let x = [1,2,3];
     /// let y = x.map(|v| v + 1);
     /// assert_eq!(y, [2,3,4]);
```
```suggestion
    /// let x = [1, 2, 3];
    /// let y = x.map(|v| v + 1);
    /// assert_eq!(y, [2, 3, 4]);
```

JulianKnodt replied:
Is this not fixed by `./x.py fmt`?

lcnr commented on library/core/src/array/mod.rs:
```diff
@@ -364,3 +364,54 @@ macro_rules! array_impl_default {
+            dst.write(f(src));
+            guard.initialized += 1;
+        }
+        // FIXME convert to crate::mem::transmute when works with generics
```
```suggestion
        // FIXME: Convert to crate::mem::transmute once it works with generics.
```
nit

lcnr commented on library/core/src/array/mod.rs:
```diff
@@ -364,3 +364,54 @@ macro_rules! array_impl_default {
+        // unsafe { crate::mem::transmute::<[MaybeUninit<U>; N], [U; N]>(dst) }
+        crate::mem::forget(guard);
+        // SAFETY: At this point we've properly initialized the whole array
+        // and we just need to cast it to the correct type
```
```suggestion
        // and we just need to cast it to the correct type.
```
nit

lcnr reviewed (approved):
The implementation looks good to me, leaving this to:

- @rust-lang/lang for the new lang item `array` which allows us to implement methods for `[T; N]`.
- @rust-lang/libs for the new unstable method `fn array::map`.

JulianKnodt committed:
Add tracking issue #75243

Add note & example about iter order

Add doc changes

Update doc comments

pickfire commented:
Looks like this may lead to `impl Iterator for [T; N]`.

pickfire reviewed (approved):
Looks good to me. I wonder if the doc changes requires doc team to review. CC @jyn514

lcnr commented:
> Looks like this may lead to `impl Iterator for [T; N]`.

I don't think we want to implement `Iterator` of arrays, but I am not really sure what you mean with "lead to" here. THis seems more closely related to `Option::map` than `Iterator`.

We do want to implement `Iterator`(#65819) and `FromIterator`(#69985) for arrays but this method is still a lot more concise than `arr.into_iter().map(func).collect::<Result<_, _>>().unwrap()`.

JulianKnodt commented:
@pickfire ah I'm not sure if `impl Iterator for [T; N]` makes sense and it's really only map that is useful. `fold` doesn't require outputting to a fixed size(vector -> scalar sort of thing) so it can just be used with an `into_iter()`, and other things such as `filter` cannot know the output size. `map` is one of the few that makes a lot of sense because if we use a normal iterator we have to `try_convert` or other shenanigans to fixed size, which is annoying, so I think the whole point of `map` is to get around that.

pickfire commented on library/core/src/array/mod.rs:
```diff
@@ -364,3 +364,54 @@ macro_rules! array_impl_default {
+#[cfg(not(bootstrap))]
+#[lang = "array"]
+impl<T, const N: usize> [T; N] {
+    /// Returns an array of the same size as self, with `f` applied to each element.
```
@GuillaumeGomez Do you think this could be improved?

scottmcm replied:
Since it's an `FnMut`, please explicitly document the order in which the closure will be called on the items, and add an example where that matters.  For example, something like
```
let x = [1, 2, 3];
let mut temp = 0;
let y = x.map(|v| { temp += 1; v + temp });
assert_eq!(y, [1, 4, 9]);
```

pickfire commented:
Well, we could still have `fold` for array without needing to `into_iter().fold()`.

JulianKnodt commented:
Mm, that makes me worried about having too many ways to do different things without much meaningful difference. While it does remove the need to call `into_iter()`, what other benefits would it add?

scottmcm commented on library/core/src/array/mod.rs:
```diff
@@ -364,3 +364,54 @@ macro_rules! array_impl_default {
+        F: FnMut(T) -> U,
+    {
+        use crate::mem::MaybeUninit;
+        struct Guard<T, const N: usize> {
```
This type is functionally equivalent to the one used in array's `into_iter()`, just adding things to the array instead of removing things from it.  Should they be unified?

JulianKnodt replied:
ah sorry which in specific are you referring to? Is this one of the PR's open for implementing `IntoIterator for [T; N]`?

scottmcm replied:
There's no trait implementation (sadly), but the iterator type itself does exist: https://doc.rust-lang.org/std/array/struct.IntoIter.html

JulianKnodt replied:
While they are similar, this one does only iterate in one direction so unifying it might take some work. Would you think it's important to do so?

scottmcm replied:
I put it as a question since I'm not sure.  My default answer is to reduce the amount of nuanced unsafe code, but if it's awkward to merge then you can probably just leave it -- unless a real libs-impl member says otherwise, of course.

JulianKnodt replied:
Mm I do feel the use case is sufficiently different from the iterator, so I'm going to leave it for now then, unless there's a reason otherwise.

LukasKalbertodt replied:
I think merging that code is too much hassle to be worth it. The iterator also uses a range (to be a doubled ended iterator) which is not required here.

scottmcm commented:
(Lang hat on, but not speaking for the team) Personally, I'm not concerned about lang items.  If they need to exist to have methods on things in the standard library, then great, they can be implemented that way.  Fine by me. (/hat)

Awesome to see this kind of method showing up.  I'm hoping we get a bunch more of them after this sets the pattern -- I made a bunch using macros (https://lib.rs/crates/arraytools) and I'd love to never use that ever again.

GuillaumeGomez commented on library/core/src/array/mod.rs:
````diff
@@ -364,3 +364,54 @@ macro_rules! array_impl_default {
+    /// Returns an array of the same size as self, with `f` applied to each element.
+    ///
+    /// # Examples
+    /// ```
````
Please add an empty doc comment line before the codeblock.

GuillaumeGomez commented on library/core/src/array/mod.rs:
```diff
@@ -364,3 +364,54 @@ macro_rules! array_impl_default {
+#[cfg(not(bootstrap))]
+#[lang = "array"]
+impl<T, const N: usize> [T; N] {
+    /// Returns an array of the same size as self, with `f` applied to each element.
```
"\`self\`"

DutchGhost commented:
> 
> 
> Looks like this may lead to `impl Iterator for [T; N]`.

If `[T; N]` implements `FromIterator`, this function is as simple as `std::array::IntoIter::new(self).map(f).collect()`. Would that an implementation that can later be used?

pickfire commented on library/core/src/array/mod.rs:
```diff
@@ -364,3 +364,61 @@ macro_rules! array_impl_default {
+#[cfg(not(bootstrap))]
+#[lang = "array"]
+impl<T, const N: usize> [T; N] {
+    /// Returns an array of the same size as `self`, with `f` applied to each element.
```
Should we change this to
```suggestion
    /// Apply function `f` to each element in the array.
```
I don't know if mentioning that the same size is useful since that is in the type signature and similar to `Iterator::map` but I think we should clarify that `f` is a function.

JulianKnodt replied:
ah I can clarify that `f` is a function, but I think it is useful to specify that it's another array returned

pickfire commented on library/core/src/array/mod.rs:
```diff
@@ -364,3 +364,61 @@ macro_rules! array_impl_default {
+#[lang = "array"]
+impl<T, const N: usize> [T; N] {
+    /// Returns an array of the same size as `self`, with `f` applied to each element.
+    /// The closure will be called on elements 0 up to N.
```
I think we should say excluding `N`, "up to" may not be that clear and may be interpreted as including.

the8472 commented:
In the special case of `size_of::<T>() == size_of::<T>() && align_of::<T>() == align_of<U>()` this could be performed in place. Since they're const this can be evaluated at compile time and only select one or the other implementation. Together with RVO that might save a lot of extra stack space on large arrays.

JulianKnodt commented:
I think that might be a good future optimizations, but for the first pass would it be important to include such an optimization?

the8472 commented:
No, probably not needed. I'll note it in the tracking ticket then.

LukasKalbertodt commented on library/core/src/array/mod.rs:
```diff
@@ -364,3 +364,61 @@ macro_rules! array_impl_default {
+#[lang = "array"]
+impl<T, const N: usize> [T; N] {
+    /// Returns an array of the same size as `self`, with function `f` applied to each element.
+    /// The closure will be called on elements 0 up to but excluding N.
```
I think last sentence is kind of redundant. "applied to each element" should be sufficient, right?

JulianKnodt replied:
Someone suggested that I specify the order, as it's a `FnMut` closure, so it does matter how the function is called. Should I try condensing the sentences?

LukasKalbertodt replied:
Ah mh. I am not quite sure if we want to guarantee order. But I guess it's fine for now. Maybe just add "... in order" to the first sentence? Is that clear enough?

scottmcm replied:
(I'm the one who suggested documenting the order)

I suggested documenting the order because if it's `FnMut`, it's likely that people will start to depend on it, and when it's not documented there are often follow-up PRs requesting that it get documented anyway (for example, https://github.com/rust-lang/rust/pull/60396 for `retain`).

For APIs where we don't want to guarantee order than I'd personally encourage just `Fn` (though that's not an official libs guideline) because people shouldn't be surprised if they use interior mutability and the behaviour goes a bit weird.  But I think this is a case where different orders don't seem interesting, so I'd just pick "we go forward" and call it done.  (I would certainly expect that this would work like `.into_iter().map(f).collect()`.)

LukasKalbertodt replied:
Yeah, you are probably right; I can't imagine what not specifying the order would buy us. So yeah, `FnMut` and guaranteeing the obvious order seems fine! We should add it as open questions to the tracking issue though.

I would still improve this sentence as mentioned above.

LukasKalbertodt commented on library/core/src/array/mod.rs:
````diff
@@ -364,3 +364,61 @@ macro_rules! array_impl_default {
+    /// # Examples
+    ///
+    /// ```
+    /// # #![feature(array_map)]
````
We don't typically hide the feature gates in doc examples, to not needlessly confuse users.

JulianKnodt replied:
Ah hiding it was also suggested by another reviewer, I can undo it though

LukasKalbertodt commented on library/core/src/array/mod.rs:
````diff
@@ -364,3 +364,61 @@ macro_rules! array_impl_default {
+    ///
+    /// ```
+    /// # #![feature(array_map)]
+    /// let x = [1, 2, 3];
+    /// let y = x.map(|v| v + 1);
+    /// assert_eq!(y, [2, 3, 4]);
+    ///
+    /// let x = [1, 2, 3];
+    /// let mut temp = 0;
+    /// let y = x.map(|v| { temp += 1; v * temp });
+    /// assert_eq!(y, [1, 4, 9]);
````
Might I suggest an example where we change the type of the array? Maybe, something like `["Ferris", "is", "cute"].map(|s| s.len())` -> `[6, 2, 4]`.

LukasKalbertodt reviewed (commented):
I think that's a very useful method and I needed this lots of times already. For one of my projects, I even implemented `map` as an extension trait simply because I needed it that often. 

While not being in the lang team, I think adding a lang item for this is fine. We also added one for `&[u8]` to add some methods just to that slice. I would even expect more array methods to be added in the future.

lcnr commented:
@LukasKalbertodt seeing that you are interested in this method, can I assign the rest to you?

As I am not a libs team member I don't think I should merge this myself :thinking: 

r? @LukasKalbertodt feel free to reassign though

JulianKnodt committed:
Rm hiding feature gate & add 1 more example

Update order docs for `map`

LukasKalbertodt commented:
@lcnr Sure thing!

JulianKnodt commented:
I've rm'd the 2nd sentence and added "in order" to the first, let me know if there are any more changes you want made

LukasKalbertodt commented on library/core/src/array/mod.rs:
```diff
@@ -364,3 +364,65 @@ macro_rules! array_impl_default {
+            }
+        }
+        let mut dst = MaybeUninit::uninit_array::<N>();
+        let mut guard: Guard<U, N> = Guard { dst: &mut dst as *mut _ as *mut U, initialized: 0 };
```
Can you use `MaybeUninit::first_ptr_mut` here?

JulianKnodt replied:
oh yea for sure, was not aware that that was a thing

LukasKalbertodt reviewed (changes requested):
One last thing I thought of: could you add a test checking that the correct number of items are dropped? Something like a custom type that decreases some global counter when dropped. Then you call `map` with a function that panics for one item in the middle, catch the panic and then check the global counter. I think I'd be more comfortable with having such a test.

JulianKnodt committed:
Add drop check test & MaybeUninit::first_ptr_mut

Also in drop check test add hacky workaround for platforms that don't support
panic=unwind

LukasKalbertodt commented on library/core/tests/array.rs:
```diff
@@ -301,3 +301,31 @@ fn array_map() {
+
+#[test]
+fn array_map_drop_safety() {
+    static mut DROPPED: usize = 0;
```
I'd prefer using a static `AtomicUsize` or even an `Rc<Cell<usize>>` in each `DropCounter` here to avoid `unsafe` blocks below. The performance here really doesn't matter and less `unsafe` is better.

LukasKalbertodt reviewed (changes requested):
Almost ready!

LukasKalbertodt commented:
Already as a heads up for the others: I intend to merge this once the comment above is resolved. If someone disagrees (due to adding the lang item, for example) please let me know!

But as far as I see it, the lang team was pinged, one lang-team member said adding a lang-item is fine, and no one disagreed. I also saw lang items being added in similar situations, too. So I guess it's fine and I hope I am not overstepping my libs line by approving this PR :yum:

scottmcm commented:
@LukasKalbertodt Unless it's something extraordinary (like a pervasive change to UB rules), I don't think there's ever a problem with adding something unstable behind a feature flag.

cuviper commented:
The only hesitation I would have is that stabilizing this method on lib's part will also commit the compiler to providing this `lang` functionality in some form (though unstable), including its dependence on const generics.

scottmcm commented:
@cuviper Totally agreed, but I think it's ok to have that conversation on the tracking issue or the stabilization PR.  It doesn't need to block the experimentation PR.

LukasKalbertodt commented:
Sounds good, hurray :)
Thanks again for the PR!

@bors r+

bors commented:
:pushpin: Commit 8e3e27049fa12067e8840da8e78bc3098a177f45 has been approved by `LukasKalbertodt`

<!-- @bors r=LukasKalbertodt 8e3e27049fa12067e8840da8e78bc3098a177f45 -->
<!-- homu: {"type":"Approved","sha":"8e3e27049fa12067e8840da8e78bc3098a177f45","approver":"LukasKalbertodt"} -->

cuviper commented on library/core/src/array/mod.rs:
```diff
@@ -364,3 +364,66 @@ macro_rules! array_impl_default {
+        crate::mem::forget(guard);
+        // SAFETY: At this point we've properly initialized the whole array
+        // and we just need to cast it to the correct type.
+        unsafe { (&mut dst as *mut _ as *mut [U; N]).read() }
```
FWIW, this could use `transmute_copy`, which basically amounts to the same pointer casting and read.

JohnTitor commented:
Failed in rollup: https://github.com/rust-lang/rust/pull/75474#issuecomment-673202885
@bors r-

JulianKnodt commented:
@LukasKalbertodt I updated this PR with a change for platforms that don't support `panic=unwind`, was hoping you might look at it, and if you know a better solution since I just mimicked a hacky solution from elsewhere.

LukasKalbertodt commented:
Oh, interesting. Well, the solution indeed looks somewhat like a hack; I expected a `cfg` attribute instead. But as the existing test also uses this hack, it's probably fine. I certainly don't have a better idea right now.

@bors try

bors commented:
:hourglass: Trying commit af32db21c8e1b1970886eb64b9955f9590e4c445 with merge c3e50bc41073ea02c3a3d33213c16142a40d1742...
<!-- homu: {"type":"TryBuildStarted","head_sha":"af32db21c8e1b1970886eb64b9955f9590e4c445","merge_sha":"c3e50bc41073ea02c3a3d33213c16142a40d1742"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/979475560), [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=35963)
Build commit: c3e50bc41073ea02c3a3d33213c16142a40d1742 (`c3e50bc41073ea02c3a3d33213c16142a40d1742`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=35963","checks-actions":"https://github.com/rust-lang-ci/rust/runs/979475560"},"merge_sha":"c3e50bc41073ea02c3a3d33213c16142a40d1742"} -->

LukasKalbertodt commented:
@bors r+

bors commented:
:pushpin: Commit af32db21c8e1b1970886eb64b9955f9590e4c445 has been approved by `LukasKalbertodt`

<!-- @bors r=LukasKalbertodt af32db21c8e1b1970886eb64b9955f9590e4c445 -->
<!-- homu: {"type":"Approved","sha":"af32db21c8e1b1970886eb64b9955f9590e4c445","approver":"LukasKalbertodt"} -->

bors commented:
:hourglass: Testing commit af32db21c8e1b1970886eb64b9955f9590e4c445 with merge a0c290f951076d781b0ff08370445517446cee05...
<!-- homu: {"type":"BuildStarted","head_sha":"af32db21c8e1b1970886eb64b9955f9590e4c445","merge_sha":"a0c290f951076d781b0ff08370445517446cee05"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/980703299), [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=35967)
Approved by: LukasKalbertodt
Pushing a0c290f951076d781b0ff08370445517446cee05 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"LukasKalbertodt","base_ref":"master","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=35967","checks-actions":"https://github.com/rust-lang-ci/rust/runs/980703299"},"merge_sha":"a0c290f951076d781b0ff08370445517446cee05"} -->

leonardo-m commented:
I can't find one use case for this in my codebase :-(

JulianKnodt commented:
Mm, one case that is particularly useful to me is in numerical computing and performing vectorized operations with fixed size arrays. In this case, it's nice to have a method that retains the size of the array but allows performing some operation on each item in the inner type.

I would keep it in mind when looking to operate on something that is a known size at compile that must be transmuted into something that is the same size. With that said, it's not useful for a lot of cases, because I imagine that if the code is not performance sensitive and/or must be dynamically allocated, it'll just use `Vec`, which this wouldn't apply to.

scottmcm commented:
@leonardo-m Given how annoying to use arrays have historically been, I'm not that surprised.

This is one of those things that gets more useful as new things that produce arrays start showing up -- for example, from the array_chunks(_mut) iterators introduced in #74373 and #75021.

leonardo-m commented:
In my Nightly-based codebase I have probably few thousand arrays, tiny ones, small ones, and larger boxed ones. And currently in about 100% of the case I don't want to create a new array from another one, but read the contents of an array or but mutate an existing array.

electroCutie commented:
> The function added is a monadic (I think?) 

It is not Monadic. That would be more like mapping from elements intro arrays and then concatenating them as the result
This appears to be a Functor
